### PR TITLE
refactor(shared-data, api): make command schema from py models, reference in protocol schema v7

### PR DESCRIPTION
**Pull request by b-cooper**
# Overview

Create a script and make task `make command_schema version=X` that will generate new a command
schema from the pydantic models of the protocol engine. Protocol Schema v7 now references commands
via a fixed reference to this command schema. Add a snapshot test to the opentrons/api project to
ensure that the command schema has not changed with out an intentional re-running of the schema
generation task.

coauthored by: @SyntaxColoring 

# Changelog

- added `command` directory to `shared-data/python/` that contains a new `generate_command_schema` function
- added `make command_schema version=X` to the shared data python module's Makefile
- replaced the hand maintained `command` enumeration in `shared-data/protocols/schemas/7-draft.json` with a static reference to the newly generated `shared-data/commands/schemas/7.json`
- add a snapshot-style regression test to the `opentrons.api` module that ensures that the newest generated schema is identical to the output calling the `generate_command_schema` function with the current pydantic models.

# Review requests

- does this feel like a workable solution for now?

# Risk assessment
medium, 7 is still in draft, but this is a far reaching change

### CODE REVIEW

File: shared-data/protocol/schemas/7-draft.json [hunk 1]
```diff
@@ -311,976 +311,9 @@
...
-                "enum": ["magneticModule/disengage"]
-              },
-              "params": {
-                "$ref": "#/definitions/moduleOnlyParams"
-              }
-            }
-          },
-
-          {
-            "description": "Temperature Module Set Target Temperature: Module will begin moving to the target temperature. This command is non-blocking, it does not delay protocol execution while approaching the target temperature.",
```
Comment (SyntaxColoring) [inline shared-data/protocol/schemas/7-draft.json:752]: Probably shouldn't happen in this PR, but we should do a side-by-side audit of docstrings to make sure we didn't lose anything meaningful in this deduplication.

File: shared-data/python/opentrons_shared_data/command/generate_command_schema.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    schema_as_dict["$id"] = f"opentronsCommandSchemaV{version}"
+    schema_as_dict["$schema"] = "http://json-schema.org/draft-07/schema#"
+    return json.dumps(schema_as_dict, indent=2)
+
+
+def main() -> int:
+    """Handler for command line invocation to generate a command schema.
+
+    :param argv: The arguments the program was invoked with; this is usually
+                 :py:obj:`sys.argv` but if you want to override that you can.
```
Comment (SyntaxColoring) [inline shared-data/python/opentrons_shared_data/command/generate_command_schema.py]:
```suggestion
```

Comment (SyntaxColoring) [reply shared-data/python/opentrons_shared_data/command/generate_command_schema.py]: Since there is no `argv` argument to this function.

File: shared-data/python/Makefile [hunk 3]
```diff
@@ -131,3 +131,8 @@ deploy: wheel
 .PHONY: test
 test:
 	$(python) -m pytest --cov=opentrons_shared_data --cov-report xml:coverage.xml tests
+
+.PHONY: command_schema
+command_schema: 
+	$(python) opentrons_shared_data/command/generate_command_schema.py $(version) > ../command/schemas/$(version).json
```
Comment (SyntaxColoring) [inline shared-data/python/Makefile]:
```suggestion
.PHONY: command-schema
command-schema: 
	$(python) opentrons_shared_data/command/generate_command_schema.py $(COMMAND_SCHEMA_VERSION) > ../command/schemas/$(COMMAND_SCHEMA_VERSION).json
```

I think [makefile convention](https://stackoverflow.com/a/62042656/497934) is for targets to be `kebab-case`, and for variables supplied from the outside to be `UPPERCASE`.

Expanding `version` to `COMMAND_SCHEMA_VERSION` is to avoid collisions with anything else. Probably unnecessary, but I guess I'm nervous because makefile variables are basically unscoped globals.

File: api/tests/opentrons/commands/test_command_schema_snapshot.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+)
+
+
+def test_command_schema_has_not_changed() -> None:
+    """Confirm that the command schema has not been changed since it was last generated. If this test fails you may want to run `make command_schema version=X` in 'shared-data/python'"""
+    newest_version = get_newest_schema_version()
+    existing_schema_string = load_schema_string(newest_version)
+    generated_schema_string = generate_command_schema(newest_version)
+
+    assert existing_schema_string == generated_schema_string
```
Comment (SyntaxColoring) [inline api/tests/opentrons/commands/test_command_schema_snapshot.py]: Some suggestions after playing around with this a little bit:

* Provide a verbose assertion message to spell things out. This will show up in the command-line output if this test fails.
* Reverse the polarity of the `==` to make the diff more intuitive (to me, anyway). That way, if you've added an extra field to a Python model that isn't in the JSON schema, it will show up as `+`, not `-`.
* Also, shorten the function docstring to match our Python line length conventions.

```suggestion
SCHEMA_CHANGED_MESSAGE = """\
The Python models for Protocol Engine commands do not match the shared JSON schema.

If this change is accidental, undo the changes to our Python models.

Or, if this change is intentional, update the shared JSON schema by running:

    make -C shared-data/python command-schema COMMAND_SCHEMA_VERSION=<version number>

...and include the updated JSON schema file in your pull request.
"""


def test_command_schema_has_not_changed() -> None:
    """Confirm that the command schema has not changed."""
    newest_version = get_newest_schema_version()
    existing_schema_string = load_schema_string(newest_version)
    generated_schema_string = generate_command_schema(newest_version)

    assert generated_schema_string == existing_schema_string, SCHEMA_CHANGED_MESSAGE

```

File: shared-data/protocol/fixtures/7/simpleV7.json [hunk 5]
```diff
@@ -1342,7 +1357,7 @@
       }
     },
     {
-      "commandType": "delay",
+      "commandType": "waitForDuration",
```
Comment (SyntaxColoring) [inline shared-data/protocol/fixtures/7/simpleV7.json:1360]: What's the deal with this change? Did this fixture fail validation without it?

If so, we might have to fix something up, because the server does intend to support `delay` currently.

File: shared-data/python/Pipfile.lock [hunk 6]
```diff
@@ -286,26 +400,43 @@
...
         "importlib-metadata": {
             "hashes": [
-                "sha256:1208431ca90a8cca1a6b8af391bb53c1a2db74e5d1cef6ddced95d4b2062edc6",
-                "sha256:ea4c597ebf37142f827b8f39299579e31685c31d3a438b59f469406afd0f2539"
+                "sha256:d5059f9f1e8e41f80e9c56c2ee58811450c31984dfa625329ffd7c0dad88a73b",
+                "sha256:d84d17e21670ec07990e1044a99efe8d615d860fd176fc29ef5c306068fda313"
             ],
             "markers": "python_version < '3.8'",
-            "version": "==4.11.3"
+            "version": "==5.1.0"
```
Comment (mcous) [inline shared-data/python/Pipfile.lock]: I'm pretty sure this upgrade to the locked version of `importlib-metadata` is causing CI failures because of very silly Python community drama in the flake8 project, whereby flake8 refuses to support newer versions of `importlib_metadata`. We need to avoid bumping this package, somehow.

If you want to have a bad time, search `importlib-metadata` in the flake8 issue tracker.

Suggestions:
1. Move script to `api` project, or
2. Lock `importlib-metadata` explicitly in `shared-data/python/Pipfile`

File: shared-data/python/Pipfile [hunk 7]
```diff
@@ -26,3 +26,4 @@ pytest-clarity = "~=1.0.0"
 
 [packages]
 opentrons-shared-data = { editable = true, path = "." }
+opentrons = { editable = true, path = "../../api" }
```
Comment (mcous) [inline shared-data/python/Pipfile]: Should this be in `dev-packages`?

File: shared-data/python/opentrons_shared_data/command/generate_command_schema.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
+import json
```
Comment (mcous) [inline shared-data/python/opentrons_shared_data/command/generate_command_schema.py]: Since this prints to stdout, I think (per comment above) this should live in the `api` package. It removes the circular dependency caused by making this a module of `opentrons_shared_data`, and sidesteps the flake8 issue mentioned above. It also makes more sense with where the test file is located

The `shared-data` Makefile can still call into the app's Python environment, and this will only become easier once #11682 is merged and we have a single Python dev environment

File: shared-data/python/opentrons_shared_data/command/generate_command_schema.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+def generate_command_schema(version: str) -> str:
+    raw_json_schema = CreateCommandUnion.schema_json()
+    schema_as_dict = json.loads(raw_json_schema)
+    schema_as_dict["$id"] = f"opentronsCommandSchemaV{version}"
+    schema_as_dict["$schema"] = "http://json-schema.org/draft-07/schema#"
+    return json.dumps(schema_as_dict, indent=2)
+
+
+def main() -> int:
```
Comment (mcous) [inline shared-data/python/opentrons_shared_data/command/generate_command_schema.py]: Can we restructure this a little, since a function that spits out the command schema would be useful in non-CLI applications?

- Export a function called `generate_command_schema` that takes a `version` argument and returns the schema string
- Move `argparse` import + logic to inside the `if __name__ == "main":` block

**Review by mcous — COMMENTED** [review]
Looks fantastic! Looking at its usage, I think we should move the `generate_command_schema` module to the `api` project to avoid useless Python environment pain and create a little bit of source-of-truth clarity

File: shared-data/protocol/fixtures/7/simpleV7.json [hunk 10]
```diff
@@ -1266,6 +1269,9 @@
       "id": "3abc123",
       "params": {
         "labwareId": "sourcePlateId",
+        "loadName": "armadillo_96_wellplate_200ul_pcr_full_skirt",
+        "namespace": "opentrons",
+        "version": 1,
```
Comment (SyntaxColoring) [inline shared-data/protocol/fixtures/7/simpleV7.json:1274]: Since this PR makes `loadLabware`, et. al. commands their full Protocol Engine variants:

Should this PR also remove the top-level `labware`, etc. keys from our v7 schema and v7 fixtures? Or do you want that to be its own PR?

File: shared-data/python/Pipfile (hunk 7 above)
Comment (SyntaxColoring) [reply shared-data/python/Pipfile]: I'd be surprised if it could be in either `[packages]` or `[dev-packages]`, actually. Doesn't this create a circular dependency between `api` and `shared-data`, either way?

File: shared-data/protocol/fixtures/7/simpleV7.json (hunk 5 above)
Comment (b-cooper) [reply shared-data/protocol/fixtures/7/simpleV7.json:1360]: That's correct, it did fail validation, and I believe that is because the server supports 'delay' at the json ingestion level, but doesn't support the bare PE command.

File: shared-data/protocol/fixtures/7/simpleV7.json (hunk 10 above)
Comment (b-cooper) [reply shared-data/protocol/fixtures/7/simpleV7.json:1274]: I'd like that change to be it's own PR, but yes that's my intent. @shlokamin and I are going to sit down and discuss the implications for step generation and ticket out the work required to get it up to v7 support.

File: api/tests/opentrons/commands/test_command_schema_snapshot.py [hunk 11]
```diff
@@ -0,0 +1,27 @@
+from opentrons_shared_data.command import load_schema_string, get_newest_schema_version
```
Comment (mcous) [inline api/tests/opentrons/commands/test_command_schema_snapshot.py]: This test should be moved to `api/tests/opentrons/protocol_engine/commands`. Confusingly enough, `opentrons/commands` is totally different; it's the legacy `Broker` command messages

**Comment by mcous** [comment]
I don't see any changes to the `Pipfile` itself, can the `Pipfile.lock` changes be removed from this PR so that `shared-data/python` CI can pass?

File: shared-data/protocol/schemas/7-draft.json (hunk 1 above)
Comment (SyntaxColoring) [reply shared-data/protocol/schemas/7-draft.json:752]: Ticketed as RSS-158.

**Review by SyntaxColoring — APPROVED** [review]
Looks great to me once we resolve the problems with environment setup.

Thanks for pushing this through!
